File: selection_sort/javascript/selection_sort.js
export default function sort(a) {
    if (!a) {
        return;
    }
    for (let i = 0; i < a.length - 1; i += 1) {
        const j = smallest(a, i);
        swap(a, i, j);
    }
}

function smallest(a, start) {
    let s = start;
    for (let i = start + 1; i < a.length; i += 1) {
        if (a[i] < a[s]) {
            s = i;
        }
    }
    return s;
}

function swap(a, i, j) {
    const t = a[i];
    a[i] = a[j];
    a[j] = t;
}

File: selection_sort/go/selection_sort.go
package selection_sort

func Sort(a []int) {
	if a == nil {
		return
	}
	for i := 0; i < len(a)-1; i++ {
		j := smallest(a, i)
		swap(a, i, j)
	}
}

func smallest(a []int, start int) int {
	s := start
	for i := start + 1; i < len(a); i++ {
		if a[i] < a[s] {
			s = i
		}
	}
	return s
}

func swap(a []int, i, j int) {
	a[i], a[j] = a[j], a[i]
}
